### PR TITLE
Extend `wasm-proc` util functionality

### PR DESCRIPTION
**Pull request by breathx**
Due to the fact that soon we will constantly use the metadata of programs, it would be useful to improve the existing tools for interaction with them and their optimization.

The purpose of this PR is to make using the tool more comfortable both personally and in CI.

- `wasm-proc` now processes through multiple targets
- `wasm-proc` now contains flags:
  - `-m` or `--meta`: forces metadata optimization for targets. Optional.
  - `-o` or `--optimize`:  forces chain optimization for targets. Optional.
  - `-p` or `--path`: specifies paths for `.wasm` file targets . You may skip this flag, but it's args are required.

Note:
If none of `-o` and `-m` specified, tool does nothing, so it becomes useless. To avoid this situation, presence or absence at the same time of these two flags is equivalent:
```
./wasm-proc -mo arg1 arg2
# equivalents
./wasm-proc arg1 arg2
```

Also:
- Removed warnings `unused variable` from `pallets/gear/lib.rs` by marking them as intentional

@gear-tech/dev

### CODE REVIEW

File: utils/wasm-proc/src/main.rs [hunk 1]
```diff
@@ -17,38 +17,36 @@
...
-
-    // Invoke optimizer for the chain
-    let mut binary_module = module.clone();
-    let binary_file_name = PathBuf::from(input).with_extension("opt.wasm");
-    utils::optimize(&mut binary_module, vec!["handle", "init"]).expect("Optimizer failed");
-    parity_wasm::serialize_to_file(binary_file_name.clone(), binary_module)
-        .expect("Serialization failed");
+/// Calls chain optimizer
+fn optimize(path: &str, mut binary_module: Module) {
+    println!("\n*** Processing chain optimization: {}", path);
```
Comment (NikVolf) [inline utils/wasm-proc/src/main.rs:28]:
```suggestion
```

or use `log::debug!`

File: utils/wasm-proc/src/main.rs [hunk 2]
```diff
@@ -17,38 +17,36 @@
...
+
+    if let Err(_) = utils::optimize(&mut binary_module, vec!["handle", "init"]) {
+        println!("Optimizer failed");
+    }
+
+    if let Err(e) = parity_wasm::serialize_to_file(binary_file_name.clone(), binary_module) {
+        println!("Serialization failed: {}", e);
+    }
 
     println!("Optimized wasm: {}", binary_file_name.to_string_lossy());
```
Comment (NikVolf) [inline utils/wasm-proc/src/main.rs:40]:
```suggestion
```

File: utils/wasm-proc/src/main.rs [hunk 3]
```diff
@@ -57,10 +55,61 @@ fn main() {
...
+    let app = App::new("wasm-proc").args(&[meta, opt, path]);
+
+    let matches = app.get_matches();
+
+    let wasm_files: Vec<&str> = matches
+        .values_of("path")
+        .expect("Path to wasm files is required")
+        .collect();
+
+    let o = matches.is_present("optimize");
```
Comment (NikVolf) [inline utils/wasm-proc/src/main.rs:100]:
```suggestion
    let do_optimize = matches.is_present("optimize");
```

File: utils/wasm-proc/src/main.rs [hunk 4]
```diff
@@ -57,10 +55,61 @@ fn main() {
...
+    let m = matches.is_present("meta");
```
Comment (NikVolf) [inline utils/wasm-proc/src/main.rs:101]:
```suggestion
    let do_meta = matches.is_present("meta");
```

File: utils/wasm-proc/src/main.rs [hunk 5]
```diff
@@ -57,10 +55,61 @@ fn main() {
...
+    for file in wasm_files {
+        if let Ok(module) = parity_wasm::deserialize_file(file) {
+            if o || !(o || m) {
+                optimize(file, module.clone());
+            }
+            if m || !(o || m) {
+                optimize_meta(file, module.clone());
+            }
+        } else {
+            println!("Failed to load wasm file: {}", file);
```
Comment (NikVolf) [inline utils/wasm-proc/src/main.rs:112]: make `main()` return Result<(), ?>

if you print errors this way, exit code of the program would be 0

File: utils/wasm-proc/src/main.rs [hunk 6]
```diff
@@ -17,38 +17,36 @@
...
     println!("Optimized wasm: {}", binary_file_name.to_string_lossy());
+}
 
-    // Invoke optimizer for the metadata
-    let mut metadata_module = module.clone();
-    let metadata_file_name = PathBuf::from(input).with_extension("meta.wasm");
-    utils::optimize(
+/// Calls metadata optimizer
+fn optimize_meta(path: &str, mut metadata_module: Module) {
+    println!("\n*** Processing metadata optimization: {}", path);
```
Comment (NikVolf) [inline utils/wasm-proc/src/main.rs:45]:
```suggestion
```

**Comment by NikVolf** [comment]
> If none of -o and -m specified, tool does nothing, so it becomes useless. To avoid this situation, presence or absence at the same time of these two flags is equivalent:

don't deliberately produce ambiguous behavior

invert flag to *skip* meta or optimisation like it is done in plasm-utils

so `wasm-proc --skip-meta file1.wasm`

and return error if it is used like that
`wasm-proc --skip-meta --skip-opt file1.wasm`

File: utils/wasm-proc/src/main.rs (hunk 5 above)
Comment (breathx) [reply utils/wasm-proc/src/main.rs:112]: Yes, nothing crushes, we have an error in console, and script continues work with other arguments. Do you want to break processing of the script if we ran into error?

Comment (NikVolf) [reply utils/wasm-proc/src/main.rs:112]: Yes, like every sensible program does

File: utils/wasm-proc/src/main.rs (hunk 3 above)
Comment (breathx) [reply utils/wasm-proc/src/main.rs]: Changed according to new (skipping) flags logic

File: utils/wasm-proc/src/main.rs (hunk 4 above)
Comment (breathx) [reply utils/wasm-proc/src/main.rs]: Same

File: utils/wasm-proc/src/main.rs (hunk 5 above)
Comment (breathx) [reply utils/wasm-proc/src/main.rs]: Done

**Review by NikVolf — APPROVED** [review]

**Comment by NikVolf** [comment]
PR description is a bit outdated now

**Review by NikVolf — APPROVED** [review]

**Pull request merged**